### PR TITLE
Move serde string deser helper to a separate crate

### PR DESCRIPTION
**Pull request by Kixunil**
This renames `CowHelper` to `DeserStrHelper`, which is clearer and moves
it to a separate crate with better documentation and tests.

Part of #7

I took a bit more time to polish it. :) Independent of #8 because it's a separate crate.

Should we reexport it in `amplify` under the old name or the new name?

### CODE REVIEW

**Comment by dr-orlovsky** [comment]
Cool, thanks! Will review today.

> Should we reexport it in amplify under the old name or the new name?

That would be nice!

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,9 @@
 # will have compiled files and executables
 /target
 
+# sub crates may contain target too because of rls
+*/target
```
Comment (dr-orlovsky) [inline .gitignore:6]: may be just `target/` than?

File: serde_str_helpers/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+[package]
+name = "serde_str_helpers"
+version = "0.1.0"
+authors = ["Martin Habovstiak <martin.habovstiak@gmail.com>", "Dr. Maxim Orlovsky <orlovsky@pandoracore.com>"]
```
Comment (dr-orlovsky) [inline serde_str_helpers/Cargo.toml:4]: I'm not an author here yet, just a reviewer :)

File: serde_str_helpers/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,122 @@
+// Rust language amplification library providing multiple generic trait
+// implementations, type wrappers, derive macros and other language enhancements
+//
+// Written in 2019-2020 by
+//     Martin Habovstiak <martin.habovstiak@gmail.com>
+//     Dr. Maxim Orlovsky <orlovsky@pandoracore.com>
```
Comment (dr-orlovsky) [inline serde_str_helpers/src/lib.rs:6]: the same as above: not yet an author :)

File: serde_str_helpers/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            fn try_from(value: DeserStrHelper<'a>) -> Result<Self, Self::Error> {
+                if let Cow::Owned(_) = value.into() {
+                    panic!("String not borrowed");
+                }
+
+                Ok(CheckDeser)
+            }
+        }
+
+        let _ = serde_json::from_str::<CheckDeser>("\"The Times 03/Jan/2009 Chancellor on brink of second bailout for banks\"").unwrap();
```
Comment (dr-orlovsky) [inline serde_str_helpers/src/lib.rs:120]: Not sure why we need double quote marks here, but it's an absolute nit

File: serde_str_helpers/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+///
+/// ## Example
+///
+/// ```
+/// use serde_derive::Deserialize;
+/// use serde_str_helpers::DeserStrHelper;
+/// use std::convert::TryFrom;
+///
+/// #[derive(Deserialize)]
+/// #[serde(try_from = "DeserStrHelper")]
```
Comment (dr-orlovsky) [inline serde_str_helpers/src/lib.rs:47]: From this sample it follows that 
```rust
#[serde(try_from = "BorrowStr")]
```
may be more nice as a naming?

**Review by dr-orlovsky — APPROVED** [review]
Only nits, otherwise LGTM & utACK

**Comment by dr-orlovsky** [comment]
It seems like `cargo fmt --all` does not allow this to pass though...

File: serde_str_helpers/src/lib.rs (hunk 4 above)
Comment (Kixunil) [reply serde_str_helpers/src/lib.rs:120]: Json serialization :)

File: serde_str_helpers/src/lib.rs (hunk 5 above)
Comment (Kixunil) [reply serde_str_helpers/src/lib.rs:47]: Hmm, good question. It looks quite worse on its own though. `DeserBorrowStr`?

**Comment by Kixunil** [comment]
> That would be nice!

So which one? :) (renaming is breaking)

I really dislike that `fmt`, how about bumping max line length?

**Comment by dr-orlovsky** [comment]
> So which one? :) (renaming is breaking)

Any that you prefer, you are the author. If you do not have a strong opinion, I think while the first one (`...Helper`) is easy to read in general, the second one (`BorrowStr`) makes much more sense when seen in the code, since it's self/explanatory what it does (`Helper` is generic and it's hard to understand in which way it helps).

Fmt: up to you. I really like this onld days linux kernel-style requirements of not having more than 80 chars per line (c-lightning is also follow them), but in rust world they are not widely used. Feel free to push a comment to #11 (you should have all permissions to push into my PRs) with rustfmt settings (and reformatting) you prefer.

**Comment by Kixunil** [comment]
I mean whether to rename it or not in `amplify`.

Regarding limit, I'm not keen to blow it up to insane values but since current screens are quite wider and some statements seem unreadable to me (especially putting a single function argument on a separate line) I'd like to see a reasonable increase. Do you think the default is too long?

**Comment by Kixunil** [comment]
I've unlinked the issue because this PR isn't sufficient to close it.

**Review by dr-orlovsky — APPROVED** [review]
Should I merge it for now or you'd like to work on it further?

**Comment by Kixunil** [comment]
If you don't mind leaving fmt unresolved, then you can merge it now.

**Comment by dr-orlovsky** [comment]
I actually can't merge, since I implemented the whole proper review process and github now prevents me from doing that :). So I need you to accept #11 (adding any fmt changes you'd like), after which these tests will succeed and I will be able to merge :D

**Comment by Kixunil** [comment]
Oh, now I noticed codecov report. That's super-strange. How can it decrease if it has more tests than the previous code?

**Comment by dr-orlovsky** [comment]
I think codecov just have some undeterminism for <1-2% estimations which may vary from build to build. Noticed that some time ago

**Comment by Kixunil** [comment]
Ah, OK, added some trivial tests anyway, but too late. :D

**Pull request merged**